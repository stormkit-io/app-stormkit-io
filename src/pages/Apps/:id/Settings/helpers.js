/**
 * Receive a repo address in Stormkit format and
 * transform it to a URL.
 */
export const toRepoAddr = repo => {
  if (!repo) {
    return;
  }

  const [provider, ...rest] = repo.split("/");
  const address = rest.join("/");

  if (provider === "bitbucket") {
    return `bitbucket.org:${address}.git`;
  }

  if (provider === "github") {
    return `github.com:${address}.git`;
  }

  if (provider === "gitlab") {
    return `gitlab.com:${address}.git`;
  }
};

/**
 * Formats the repository address in a Stormkit format.
 */
export const formatRepo = repo => {
  if (!repo) {
    return;
  }

  const providers = ["github.com", "bitbucket.org", "gitlab.com"];

  for (let i = 0; i < providers.length; i++) {
    const [provider] = providers[i].split(".");

    if (repo.indexOf(providers[i]) > -1) {
      if (repo.indexOf(`${providers[i]}:`) > -1) {
        // git@github.com:stormkit-io/app-www.git
        const [, project = ""] = repo.split(":");
        return `${provider}/${project.replace(".git", "")}`;
      } else {
        // https://github.com/stormkit-io/app-www
        const [, project] = repo.split(`${providers[i]}/`);
        return `${provider}/${project}`;
      }
    }
  }
};

/**
 * Prefixes the hashtag (#) to the channel name to follow the Slack naming conventions.
 */
export const formattedSlackChannelName = channel => {
  if (!channel) {
    return;
  }

  return `#${channel.replace(/^#/, "")}`;
};
