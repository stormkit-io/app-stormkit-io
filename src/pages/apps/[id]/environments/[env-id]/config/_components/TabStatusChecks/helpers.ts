export const buildStatusChecks = (
  existingChecks: StatusCheck[],
  modifiedCheck: StatusCheck | undefined,
  index: number
): StatusCheck[] => {
  const copy = [...existingChecks];

  if (typeof modifiedCheck === "undefined") {
    copy.splice(index, 1);
    return copy;
  }

  if (index > -1) {
    copy[index] = modifiedCheck;
  } else {
    copy.push(modifiedCheck);
  }

  if (new Set(copy.map(i => i.cmd.trim())).size !== copy.length) {
    throw new Error("duplicate");
  }

  return copy;
};
