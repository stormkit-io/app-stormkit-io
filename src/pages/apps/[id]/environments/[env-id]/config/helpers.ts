export const isFrameworkRecognized = (framework?: string): boolean => {
  if (!framework) {
    return false;
  }

  return ["nuxt", "next"].includes(framework);
};
