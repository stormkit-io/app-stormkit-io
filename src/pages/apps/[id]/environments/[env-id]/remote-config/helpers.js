export const sortConfigByKey = (config) => {
  const ordered = {};

  Object.keys(config)
    .sort()
    .forEach((key) => {
      ordered[key] = config[key];
    });

  return ordered;
};

export const keyToName = {
  appVersion: "App version",
  segment: "Segment",
  percentile: "Percentile",
  value: "Value",
};
