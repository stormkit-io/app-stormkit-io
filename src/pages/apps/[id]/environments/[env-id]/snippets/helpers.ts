export const isUndef = (a: unknown): boolean => typeof a === "undefined";

export const normalize = (snippets: Snippets): Snippets => {
  const head = snippets.head || [];
  const body = snippets.body || [];

  return {
    head: head.map((s: Snippet, _i: number) => ({
      ...s,
      _injectLocation: "head",
      _i
    })),
    body: body.map((s: Snippet, _i: number) => ({
      ...s,
      _injectLocation: "body",
      _i
    }))
  };
};
