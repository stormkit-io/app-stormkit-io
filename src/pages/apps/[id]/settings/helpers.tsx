/**
 * Receive a repo address in Stormkit format and
 * transform it to a URL.
 */
export const toRepoAddr = (repo: string): string => {
  if (!repo) {
    return "";
  }

  const [provider, ...rest] = repo.split("/");
  const address = rest.join("/");

  if (provider === "bitbucket") {
    return `bitbucket.org:${address}.git`;
  }

  if (provider === "github") {
    return `github.com:${address}.git`;
  }

  if (provider === "gitlab") {
    return `gitlab.com:${address}.git`;
  }

  return "";
};

/**
 * Formats the repository address in a Stormkit format.
 */
export const formatRepo = (repo: string): string => {
  if (!repo) {
    return "";
  }

  const providers = ["github.com", "bitbucket.org", "gitlab.com"];

  for (let i = 0; i < providers.length; i++) {
    const [provider] = providers[i].split(".");

    if (repo.indexOf(providers[i]) > -1) {
      if (repo.indexOf(`${providers[i]}:`) > -1) {
        // git@github.com:stormkit-io/app-www.git
        const [, project = ""] = repo.split(":");
        return `${provider}/${project.replace(".git", "")}`;
      } else {
        // https://github.com/stormkit-io/app-www
        const [, project] = repo.split(`${providers[i]}/`);
        return `${provider}/${project}`;
      }
    }
  }

  return "";
};

/**
 * Prefixes the hashtag (#) to the channel name to follow the Slack naming conventions.
 */
export const formattedSlackChannelName = (channel: string): string => {
  if (!channel) {
    return "";
  }

  return `#${channel.replace(/^#/, "")}`;
};
