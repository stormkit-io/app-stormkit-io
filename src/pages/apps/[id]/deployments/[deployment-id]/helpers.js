import React from "react";
import Link from "~/components/Link";
import { formattedDate } from "~/utils/helpers/deployments";
import ExitStatus from "../_components/ExitStatus";

/**
 * Based on the deploy log step status, it returns an exit code. If the
 * deployment is still running and this is the last step, it returns null
 * which translates to running state. Otherwise returns 0 for success and
 * 1 for failed deployments (similar to bash).
 */
export const getExitCode = ({ deploy, status, index }) => {
  const isLast = index === deploy.logs.length - 1;
  const exitCodeSuccess = 0;
  const exitCodeFailure = 1;

  if (isLast && deploy.isRunning) {
    return null;
  }

  return status ? exitCodeSuccess : exitCodeFailure;
};

/**
 * Based on the deployment and commit information prepares an array
 * that can be iterated to display the deployment steps.
 */
export const prepareSettings = ({ deploy, commit }) => [
  {
    text: "Branch",
    value: commit.branch && (
      <span>
        <span className="fas fa-code-branch mr-1" />
        {commit.branch}
      </span>
    )
  },
  { text: "Configuration", value: deploy.config?.env },
  {
    text: "Started",
    value: deploy.createdAt && formattedDate(deploy.createdAt)
  },
  {
    text: "Stopped",
    value: deploy.stoppedAt && formattedDate(deploy.stoppedAt)
  },
  { text: "Build Status", value: <ExitStatus code={deploy.exit} /> },
  {
    text: "Endpoint",
    value: deploy.exit === 0 && deploy.preview && (
      <Link to={deploy.preview} secondary>
        {deploy.preview}
        <i className="fas fa-external-link-square-alt ml-2" />
      </Link>
    )
  }
];
