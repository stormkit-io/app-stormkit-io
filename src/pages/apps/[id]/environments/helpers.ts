import { toArray } from "~/utils/helpers/array";

export const prepareBuildObject = (
  values: Record<string, string>,
  isServerless: boolean
): BuildConfig => {
  const vars: Record<string, string> = {};

  const keys = toArray<string>(values["build.vars.keys"]);
  const vals = toArray<string>(values["build.vars.values"]);

  keys.forEach((key, i) => {
    vars[key] = vals[i].replace(/^['"]+|['"]+$/g, "");
  });

  const build: BuildConfig = {
    cmd: values["build.cmd"],
    entry: values["build.entry"] || "",
    distFolder: values["build.distFolder"] || "",
    vars
  };

  if (!build.cmd) {
    build.cmd = "echo 'skip build step'";
  }

  if (isServerless && !build.entry) {
    build.entry = "__SSR__";
  }

  return build;
};
