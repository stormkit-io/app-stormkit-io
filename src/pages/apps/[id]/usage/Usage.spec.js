import router from "react-router";
import { fireEvent, waitFor } from "@testing-library/react";
import { withMockContext, waitForPromises } from "~/testing/helpers";
import { mockApp, mockUserResponse } from "~/testing/data";
import { mockFetchUsage } from "~/testing/nocks";

const fileName = "pages/apps/[id]/usage/Usage";

describe(fileName, () => {
  let wrapper;
  let app;
  let user;

  describe("api success", () => {
    beforeEach(() => {
      app = mockApp();
      user = mockUserResponse().user;
      mockFetchUsage({ app });
      wrapper = withMockContext({
        path: `~/${fileName}`,
        props: {
          app,
          user,
        },
      });
    });

    test.skip("should display a spinner", () => {
      expect(wrapper.container).toMatchSnapshot();
    });

    test.skip("should list number of deployments", async () => {
      await waitFor(() => {
        expect(wrapper.getByText("Number of Deployments")).toBeTruthy();
        expect(wrapper.getByText(/3\s\/\s15/)).toBeTruthy();
      });
    });

    test.skip("should show upgrade account for non-enterprise users", async () => {
      await waitFor(() => {
        expect(wrapper.getByText(/Upgrade account/)).toBeTruthy();
      });
    });
  });

  describe("api error", () => {
    beforeEach(() => {
      app = mockApp();
      user = mockUserResponse().user;
      mockFetchUsage({ app, status: 401 });
      wrapper = withMockContext({
        path: `~/${fileName}`,
        props: {
          app,
          user,
        },
      });
    });

    test.skip("should show an error message", async () => {
      await waitFor(() => {
        expect(
          wrapper.getByText("Something went wrong while fetching usage data.")
        ).toBeTruthy();
      });
    });
  });

  describe("enterprise users", () => {
    beforeEach(() => {
      app = mockApp();
      user = mockUserResponse().user;
      user.package.id = "enterprise";
      mockFetchUsage({ app });
      wrapper = withMockContext({
        path: `~/${fileName}`,
        props: {
          app,
          user,
        },
      });
    });

    test.skip("should show upgrade account for non-enterprise users", async () => {
      await waitFor(() => {
        expect(wrapper.getByText("Number of Deployments")).toBeTruthy();
        expect(() => wrapper.getByText(/Upgrade account/)).toThrow();
      });
    });
  });
});
