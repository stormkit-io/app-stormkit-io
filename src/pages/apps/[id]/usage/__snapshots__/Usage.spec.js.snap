// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/apps/[id]/usage/Usage api success should display a spinner 1`] = `
<div>
  <div>
    <h1
      class="mb-4 text-2xl text-white"
    >
      Usage
    </h1>
    <div
      class="spinner rounded-full w-10 h-10 bg-pink-50"
    >
      <div
        class="relative w-full h-full"
      >
        <div
          class="spinner-bounce w-full h-full rounded-full bg-white absolute top-0 left-0"
        />
        <div
          class="spinner-bounce w-full h-full rounded-full bg-white absolute top-0 left-0"
        />
      </div>
    </div>
  </div>
</div>
`;
